File: read_keys_SH3.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <time.h>

#define PINCTRL_REG_BASE 0x80018000

#define PINCTRL_BANK2_PULLUP_REG 0x620
#define PINCTRL_BANK4_PULLUP_REG 0x640

#define PINCTRL_DRIVE10_REG 0x3A0
#define PINCTRL_DRIVE17_REG 0x410

#define PINCTRL_BANK2_OUTPUT_REG 0x720
#define PINCTRL_BANK4_OUTPUT_REG 0x740

#define PINCTRL_BANK2_OUTPUT_ENABLE_REG 0xB20
#define PINCTRL_BANK4_OUTPUT_ENABLE_REG 0xB40

#define PINCTRL_BANK4_INPUT_REG 0x940

#define REG_SET 0x4
#define REG_CLR 0x8

void keyread(void *mapped_addr, char *keys) {
    unsigned int in;
    struct timespec wait = {0, 100000};
    // Bank2 Pin16~21
    for (int i = 0; i < 7; i++) {
        *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK2_OUTPUT_REG + REG_SET) = 0x10000 << i;
        nanosleep(&wait, NULL);
        in = *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_INPUT_REG);
        in = ~((in ^ (in >> 1)) & 0x1f ^ (in >> 1)) & 0x7f;
        *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK2_OUTPUT_REG + REG_CLR) = 0x10000 << i;
        nanosleep(&wait, NULL);
        keys[i] = (char)in;
    }
    // Bank4 Pin8
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_OUTPUT_REG + REG_SET) = 0x100;
    nanosleep(&wait, NULL);
    in = *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_INPUT_REG);
    in = ~((in ^ (in >> 1)) & 0x1f ^ (in >> 1)) & 0x7f;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_OUTPUT_REG + REG_CLR) = 0x100;
    nanosleep(&wait, NULL);
    keys[6] = (char)in;
}

int main() {
    /* init */
    int fd = open("/dev/mem", O_RDWR | O_SYNC);
    if (fd < 0) {
        printf("failed to open /dev/mem");
        return -1;
    }

    void *mapped_addr = mmap(NULL, 4096, PROT_READ | PROT_WRITE, MAP_SHARED, fd, PINCTRL_REG_BASE);
    if (mapped_addr == MAP_FAILED) {
        printf("failed to mapping addr");
        close(fd);
        return -1;
    }

    /* GPIO setup */
    *(volatile unsigned int *)(mapped_addr + PINCTRL_DRIVE10_REG + REG_SET) = 0x00444444;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_DRIVE17_REG + REG_SET) = 0x44444444;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK2_PULLUP_REG + REG_SET) = 0x003F03FF;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK2_OUTPUT_REG + REG_SET) = 0x00000000;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_OUTPUT_REG + REG_SET) = 0x00000000;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK2_OUTPUT_ENABLE_REG + REG_SET) =
        0x0F3FF5FF;
    *(volatile unsigned int *)(mapped_addr + PINCTRL_BANK4_OUTPUT_ENABLE_REG + REG_SET) =
        0x0001FF20;

    /* read */
    char keys[7], oldkeys[7] = {0};
    while (1) {
        keyread(mapped_addr, keys);
        if (memcmp(keys, oldkeys, sizeof(keys))) {
            memcpy(oldkeys, keys, sizeof(keys));
            continue;
        }
        unsigned char temp = 0;
        for (int i = 0; i < 7; i++) {
            temp |= keys[i];
        }
        if (temp == 0)
            continue;
        // key decode
        if (temp & 0x1)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x1) == 0) {
                    if (i == 0)
                        printf("英和和英 ");
                    if (i == 1)
                        printf("調べる ");
                    if (i == 2)
                        printf("国語漢字 ");
                    if (i == 3)
                        printf("My辞書 ");
                    if (i == 4)
                        printf("履歴/しおり ");
                    if (i == 5)
                        printf("マーカーテスト ");
                    if (i == 6)
                        printf("HOME ");
                }
        if (temp & 0x2)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x2) == 0) {
                    if (i == 0)
                        printf("Q ");
                    if (i == 1)
                        printf("W ");
                    if (i == 2)
                        printf("E ");
                    if (i == 3)
                        printf("T ");
                    if (i == 4)
                        printf("Y ");
                    if (i == 5)
                        printf("U ");
                    if (i == 6)
                        printf("I ");
                }
        if (temp & 0x4)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x4) == 0) {
                    if (i == 0)
                        printf("A ");
                    if (i == 1)
                        printf("S ");
                    if (i == 2)
                        printf("D ");
                    if (i == 3)
                        printf("G ");
                    if (i == 4)
                        printf("H ");
                    if (i == 5)
                        printf("O ");
                    if (i == 6)
                        printf("P ");
                }
        if (temp & 0x8)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x8) == 0) {
                    if (i == 0)
                        printf("Z ");
                    if (i == 1)
                        printf("X ");
                    if (i == 2)
                        printf("C ");
                    if (i == 3)
                        printf("B ");
                    if (i == 4)
                        printf("J ");
                    if (i == 5)
                        printf("K ");
                    if (i == 6)
                        printf("L ");
                }
        if (temp & 0x10)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x10) == 0) {
                    if (i == 0)
                        printf("Shift ");
                    if (i == 1)
                        printf("R ");
                    if (i == 2)
                        printf("F ");
                    if (i == 3)
                        printf("記号 ");
                    if (i == 4)
                        printf("N ");
                    if (i == 5)
                        printf("M ");
                    if (i == 6)
                        printf("- ");
                }
        if (temp & 0x20)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x20) == 0) {
                    if (i == 0)
                        printf("V ");
                    if (i == 1)
                        printf("PageUp ");
                    if (i == 2)
                        printf("Space ");
                    if (i == 3)
                        printf("Left ");
                    if (i == 4)
                        printf("Up ");
                    if (i == 5)
                        printf("Down ");
                    if (i == 6)
                        printf("Right ");
                }
        if (temp & 0x40)
            for (int i = 0; i < 7; i++)
                if ((keys[i] & 0x40) == 0) {
                    if (i == 0)
                        printf("文字切替 ");
                    if (i == 1)
                        printf("PageDown ");
                    if (i == 2)
                        printf("ツール ");
                    if (i == 3)
                        printf("Esc ");
                    if (i == 4)
                        printf("Enter ");
                    if (i == 5)
                        printf("? ");
                    if (i == 6)
                        printf("BackSpace ");
                }
        printf("\nkeys = 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x\n", keys[0], keys[1],
               keys[2], keys[3], keys[4], keys[5], keys[6]);
        memcpy(oldkeys, keys, sizeof(keys));
    }

    /* deinit */
    munmap(mapped_addr, 4096);
    close(fd);
    return 0;
}

File: read_keys_SH3.py
import gpio  # https://github.com/vitiral/gpio
import time


def setup():
    # Bank2 Pin16~PIN21
    gpio.setup(80, gpio.OUT)
    gpio.setup(81, gpio.OUT)
    gpio.setup(82, gpio.OUT)
    gpio.setup(83, gpio.OUT)
    gpio.setup(84, gpio.OUT)
    gpio.setup(85, gpio.OUT)
    # Bank4 Pin0~7
    gpio.setup(128, gpio.IN)
    gpio.setup(129, gpio.IN)
    gpio.setup(130, gpio.IN)
    gpio.setup(131, gpio.IN)
    gpio.setup(132, gpio.IN)
    gpio.setup(133, gpio.IN)
    gpio.setup(134, gpio.IN)
    gpio.setup(135, gpio.IN)
    # Bank4 Pin8
    gpio.setup(136, gpio.OUT)


def gpioread():
    keys = []
    for writeport in range(80, 87):
        if writeport == 86:
            writeport = 136
        gpio.set(writeport, 1)
        time.sleep(0.002)
        key = 0
        for readport in range(135, 127, -1):
            key = (key << 1) + gpio.read(readport)
        key = ~((key ^ (key >> 1)) & 0x1f ^ (key >> 1)) & 0x7f
        keys.append(key)
        gpio.set(writeport, 0)
    return keys


def keyread():
    keys = readgpio()
    verify_keys = readgpio()
    if not keys == verify_keys:
        return
    temp = 0
    for key in keys:
        temp = temp | key
    if temp & 0x1:
        for i, key in enumerate(keys):
            if key & 0x1 == 0:
                if i == 0:
                    print("英和和英", end=" ")
                if i == 1:
                    print("調べる", end=" ")
                if i == 2:
                    print("国語漢字", end=" ")
                if i == 3:
                    print("My辞書", end=" ")
                if i == 4:
                    print("履歴/しおり", end=" ")
                if i == 5:
                    print("マーカーテスト", end=" ")
                if i == 6:
                    print("HOME", end=" ")
    if temp & 0x2:
        for i, key in enumerate(keys):
            if key & 0x2 == 0:
                if i == 0:
                    print("Q", end=" ")
                if i == 1:
                    print("W", end=" ")
                if i == 2:
                    print("E", end=" ")
                if i == 3:
                    print("T", end=" ")
                if i == 4:
                    print("Y", end=" ")
                if i == 5:
                    print("U", end=" ")
                if i == 6:
                    print("I", end=" ")
    if temp & 0x4:
        for i, key in enumerate(keys):
            if key & 0x4 == 0:
                if i == 0:
                    print("A", end=" ")
                if i == 1:
                    print("S", end=" ")
                if i == 2:
                    print("D", end=" ")
                if i == 3:
                    print("G", end=" ")
                if i == 4:
                    print("H", end=" ")
                if i == 5:
                    print("O", end=" ")
                if i == 6:
                    print("P", end=" ")
    if temp & 0x8:
        for i, key in enumerate(keys):
            if key & 0x8 == 0:
                if i == 0:
                    print("Z", end=" ")
                if i == 1:
                    print("X", end=" ")
                if i == 2:
                    print("C", end=" ")
                if i == 3:
                    print("B", end=" ")
                if i == 4:
                    print("J", end=" ")
                if i == 5:
                    print("K", end=" ")
                if i == 6:
                    print("L", end=" ")
    if temp & 0x10:
        for i, key in enumerate(keys):
            if key & 0x10 == 0:
                if i == 0:
                    print("Shift", end=" ")
                if i == 1:
                    print("R", end=" ")
                if i == 2:
                    print("F", end=" ")
                if i == 3:
                    print("記号", end=" ")
                if i == 4:
                    print("N", end=" ")
                if i == 5:
                    print("M", end=" ")
                if i == 6:
                    print("-", end=" ")
    if temp & 0x20:
        for i, key in enumerate(keys):
            if key & 0x20 == 0:
                if i == 0:
                    print("V", end=" ")
                if i == 1:
                    print("PageUp", end=" ")
                if i == 2:
                    print("Space", end=" ")
                if i == 3:
                    print("Left", end=" ")
                if i == 4:
                    print("Up", end=" ")
                if i == 5:
                    print("Down", end=" ")
                if i == 6:
                    print("Right", end=" ")
    if temp & 0x40:
        for i, key in enumerate(keys):
            if key & 0x40 == 0:
                if i == 0:
                    print("文字切替", end=" ")
                if i == 1:
                    print("PageDown", end=" ")
                if i == 2:
                    print("ツール", end=" ")
                if i == 3:
                    print("Esc", end=" ")
                if i == 4:
                    print("Enter", end=" ")
                if i == 5:
                    print("?", end=" ")
                if i == 6:
                    print("BackSpace", end=" ")
    keys = [f"{key:#04x}" for key in keys]

    if not temp == 0:
        print(f"\nKeys = {', '.join(keys)}")


setup()
while True:
    time.sleep(0.001)
    keyread()
